Unable to push publish plugins

**Describe the bug**
Attempting to push publish plugins results in an error attempting to generate a directory

```
[20/01/22 14:43:49:427 UTC]  INFO publishing.PublisherAPIImpl: Start of Bundler: OSGIBundler
[20/01/22 14:43:49:445 UTC] ERROR publishing.PublisherAPIImpl: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
com.dotcms.publishing.DotBundleException: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:119) ~[ee-22.01.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:118) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl$AjcClosure1.run(PublisherAPIImpl.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:47) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute_aroundBody0(PublisherQueueJob.java:186) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob$AjcClosure1.run(PublisherQueueJob.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute(PublisherQueueJob.java:112) ~[dotcms_22.01_999999.jar:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:223) ~[dot.quartz-all-1.8.6_2.jar:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549) ~[dot.quartz-all-1.8.6_2.jar:?]
Caused by: java.io.IOException: Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFile(FileUtils.java:777) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:706) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:102) ~[ee-22.01.jar:?]
	... 17 more
[20/01/22 14:43:49:445 UTC] ERROR business.PublisherQueueJob: Unable to publish Bundle '01FSVZ34DPME6FZT8ZEE90KSXC': com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
com.dotcms.publishing.DotPublishingException: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:154) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl$AjcClosure1.run(PublisherAPIImpl.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:47) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish(PublisherAPIImpl.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute_aroundBody0(PublisherQueueJob.java:186) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob$AjcClosure1.run(PublisherQueueJob.java:1) ~[dotcms_22.01_999999.jar:?]
	at org.aspectj.runtime.reflect.JoinPointImpl.proceed(JoinPointImpl.java:149) ~[aspectjrt-1.9.2.jar:?]
	at com.dotcms.aspects.aspectj.AspectJDelegateMethodInvocation.proceed(AspectJDelegateMethodInvocation.java:42) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.interceptors.CloseDBIfOpenedMethodInterceptor.invoke(CloseDBIfOpenedMethodInterceptor.java:29) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.aspects.aspectj.CloseDBIfOpenedAspect.invoke(CloseDBIfOpenedAspect.java:41) ~[dotcms_22.01_999999.jar:?]
	at com.dotcms.publisher.business.PublisherQueueJob.execute(PublisherQueueJob.java:112) ~[dotcms_22.01_999999.jar:?]
	at org.quartz.core.JobRunShell.run(JobRunShell.java:223) ~[dot.quartz-all-1.8.6_2.jar:?]
	at org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:549) ~[dot.quartz-all-1.8.6_2.jar:?]
Caused by: com.dotcms.publishing.DotBundleException: com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler : generate()Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created: Unable to pull content
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:119) ~[ee-22.01.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:118) ~[dotcms_22.01_999999.jar:?]
	... 16 more
Caused by: java.io.IOException: Destination '/osgiBundles/com.dotcms.rest-osgi-debug-0.2.jar' directory cannot be created
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFile(FileUtils.java:777) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.repackage.org.apache.commons.io.FileUtils.copyFileToDirectory(FileUtils.java:706) ~[dot.commons-io-2.0.1_2.jar:2.0.1]
	at com.dotcms.enterprise.publishing.remote.bundler.OSGIBundler.generate(SourceFile:102) ~[ee-22.01.jar:?]
	at com.dotcms.publishing.PublisherAPIImpl.publish_aroundBody0(PublisherAPIImpl.java:118) ~[dotcms_22.01_999999.jar:?]
	... 16 more
[20/01/22 14:43:49:445 UTC]  INFO util.PushPublishLogger: class com.dotcms.publisher.business.PublisherQueueJob : Status Update: Failed to bundle '01FSVZ34DPME6FZT8ZEE90KSXC'
```

Reproduced on: 21.06.6, 22.01
Related Ticket: https://dotcms.zendesk.com/agent/tickets/106313

**To Reproduce**
1. Create a sender/receiver environment
2. Add a plugin to the sender
3. Attempt to push publish that plugin

**Expected behavior**
The plugin should push successfully.

discussion https://github.com/dotCMS/core/discussions/21590

--- a/dotCMS/src/main/java/com/dotcms/integritycheckers/HostIntegrityChecker.java
+++ b/dotCMS/src/main/java/com/dotcms/integritycheckers/HostIntegrityChecker.java
@@ -359,7 +359,7 @@ public class HostIntegrityChecker extends AbstractIntegrityChecker {
 
     /**
      * Perform actual fix in real tables based on hosts_ir table record.
-     *1
+     *
      * @param row map representing a row from hosts_ir table
      * @param isLastConflict flag telling  if this is the last conflict
      * @throws DotDataException

--- a/dotCMS/src/main/java/com/dotcms/publisher/business/PublishAuditHistory.java
+++ b/dotCMS/src/main/java/com/dotcms/publisher/business/PublishAuditHistory.java
@@ -12,7 +12,7 @@ import java.util.Map;
 public class PublishAuditHistory implements Serializable {
 	private static final long serialVersionUID = 1L;
 	
-	//This map contains the enpoints group
+	//This map contains the endpoints group
 	//each group can have one or more endpoints
 	private Map<String, Map<String, EndpointDetail>> endpointsMap;
 	
@@ -115,8 +115,8 @@ public class PublishAuditHistory implements Serializable {
 	
 	static XStream xstream=new XStream(new DomDriver());
 	public String getSerialized() {
-		
-	       
+
+
         String xml=xstream.toXML(this);
         return xml;
 	}

--- a/dotCMS/src/main/java/com/dotcms/publisher/pusher/wrapper/OSGIWrapper.java
+++ b/dotCMS/src/main/java/com/dotcms/publisher/pusher/wrapper/OSGIWrapper.java
@@ -1,6 +1,9 @@
 package com.dotcms.publisher.pusher.wrapper;
 
 import com.dotcms.publishing.PublisherConfig.Operation;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+
 
 /**
  * @author Jonathan Gamba
@@ -11,7 +14,8 @@ public class OSGIWrapper {
     private Operation operation;
     private String jarName;
 
-    public OSGIWrapper ( String jarName, Operation operation ) {
+    @JsonCreator
+    public OSGIWrapper (@JsonProperty("jarName") String jarName, @JsonProperty("operation") Operation operation ) {
         this.operation = operation;
         this.jarName = jarName;
     }

--- a/dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java
+++ b/dotCMS/src/main/java/com/dotcms/publisher/util/dependencies/DependencyManager.java
@@ -3,16 +3,12 @@ package com.dotcms.publisher.util.dependencies;
 import static com.dotcms.util.CollectionsUtils.set;
 
 import com.dotcms.contenttype.model.type.ContentType;
-import com.dotcms.contenttype.transform.contenttype.StructureTransformer;
 import com.dotcms.exception.ExceptionUtil;
-import com.dotcms.languagevariable.business.LanguageVariableAPI;
 import com.dotcms.publisher.business.PublishQueueElement;
-import com.dotcms.publisher.environment.bean.Environment;
 import com.dotcms.publisher.pusher.PushPublisherConfig;
 import com.dotcms.publisher.util.PublisherUtil;
 import com.dotcms.publisher.util.PusheableAsset;
 import com.dotcms.publishing.DotBundleException;
-import com.dotcms.publishing.PublisherConfig.Operation;
 import com.dotcms.publishing.PublisherFilter;
 import com.dotcms.publishing.manifest.CSVManifestBuilder;
 import com.dotcms.publishing.manifest.ManifestReason;
@@ -20,50 +16,34 @@ import com.dotmarketing.beans.Host;
 import com.dotmarketing.beans.Identifier;
 import com.dotmarketing.business.APILocator;
 import com.dotmarketing.business.CacheLocator;
-import com.dotmarketing.business.DotIdentifierStateException;
-import com.dotmarketing.business.DotStateException;
-import com.dotmarketing.business.IdentifierAPI;
-import com.dotmarketing.business.PermissionAPI;
-import com.dotmarketing.cache.FieldsCache;
 import com.dotmarketing.exception.DotDataException;
-import com.dotmarketing.exception.DotRuntimeException;
 import com.dotmarketing.exception.DotSecurityException;
-import com.dotmarketing.factories.PersonalizedContentlet;
+import com.dotmarketing.plugin.model.Plugin;
 import com.dotmarketing.portlets.categories.model.Category;
 import com.dotmarketing.portlets.containers.model.Container;
-import com.dotmarketing.portlets.containers.model.FileAssetContainer;
-import com.dotmarketing.portlets.contentlet.business.ContentletAPI;
-import com.dotmarketing.portlets.contentlet.business.HostAPI;
 import com.dotmarketing.portlets.contentlet.model.Contentlet;
-import com.dotmarketing.portlets.folders.business.FolderAPI;
 import com.dotmarketing.portlets.folders.model.Folder;
-import com.dotmarketing.portlets.htmlpageasset.model.IHTMLPage;
 import com.dotmarketing.portlets.languagesmanager.model.Language;
 import com.dotmarketing.portlets.links.model.Link;
 import com.dotmarketing.portlets.rules.model.Rule;
-import com.dotmarketing.portlets.structure.factories.StructureFactory;
-import com.dotmarketing.portlets.structure.model.Field;
 import com.dotmarketing.portlets.structure.model.Relationship;
 import com.dotmarketing.portlets.structure.model.Structure;
 import com.dotmarketing.portlets.templates.model.FileAssetTemplate;
 import com.dotmarketing.portlets.templates.model.Template;
 import com.dotmarketing.portlets.workflows.model.WorkflowScheme;
-import com.dotmarketing.util.Config;
 import com.dotmarketing.util.InodeUtils;
 import com.dotmarketing.util.Logger;
 import com.dotmarketing.util.UtilMethods;
 import com.google.common.annotations.VisibleForTesting;
-import com.google.common.collect.Table;
 import com.liferay.portal.model.User;
-import com.liferay.util.StringPool;
-import io.vavr.control.Try;
+
+import java.io.File;
+import java.nio.file.Paths;
 import java.util.*;
 import java.util.concurrent.*;
-import java.util.function.Function;
-import java.util.function.Predicate;
-import java.util.stream.Collectors;
-import java.util.stream.Stream;
+
 import org.apache.commons.lang.StringUtils;
+import org.apache.felix.framework.OSGIUtil;
 
 /**
  * The main purpose of this class is to determine all possible content
@@ -312,6 +292,21 @@ public class DependencyManager {
 							: "N/A")
 							+ " is not present in the database, not Pushed.");
 				}
+			} else if (asset.getType().equals(PusheableAsset.OSGI.getType())) {
+				final String felixLoadFolder = OSGIUtil.getInstance().getFelixDeployPath();
+				final File bundleJar = Paths.get(felixLoadFolder, asset.getAsset()).toFile();
+
+				if (bundleJar.exists()) {
+					final Plugin plugin = new Plugin();
+					plugin.setId(asset.getAsset());
+					add(plugin, PusheableAsset.OSGI);
+				} else {
+					Logger.warn(
+							this,
+							String.format(
+									"Plugin id: %s is not present as a jar, not pushed.",
+									asset.getAsset() != null ? asset.getAsset() : "N/A"));
+				}
 			}
 		}
 
@@ -423,7 +418,10 @@ public class DependencyManager {
 		} else if (Language.class.isInstance(asset)) {
 			final Language languege = Language.class.cast(asset);
 			return String.valueOf(languege.getId());
-		} else {
+		} else if (Plugin.class.isInstance(asset)) {
+			final Plugin plugin = Plugin.class.cast(asset);
+			return plugin.getId();
+		}else {
 			throw new IllegalArgumentException("Not allowed: " + asset.getClass().getName());
 		}
 	}

--- a/dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java
+++ b/dotCMS/src/main/java/com/dotcms/publishing/BundlerUtil.java
@@ -23,9 +23,14 @@ import com.dotmarketing.util.XMLUtils;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.thoughtworks.xstream.XStream;
 import com.thoughtworks.xstream.io.HierarchicalStreamWriter;
-import com.thoughtworks.xstream.io.xml.DomDriver;
 
-import java.io.*;
+import java.io.BufferedInputStream;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
 import java.nio.file.Files;
 import java.util.Calendar;
 import java.util.Date;
@@ -33,7 +38,9 @@ import java.util.List;
 
 public class BundlerUtil {
 
-    static XStream xmlSerializer = null;
+    private static ObjectMapper objectMapper;
+    private static ObjectMapper customMapper;
+
     public static final List<Status> STATUS_TO_RETRY = list(
             Status.FAILED_TO_PUBLISH, Status.SUCCESS, Status.SUCCESS_WITH_WARNINGS,
             Status.FAILED_TO_SEND_TO_ALL_GROUPS, Status.FAILED_TO_SEND_TO_SOME_GROUPS, Status.FAILED_TO_SENT
@@ -231,10 +238,6 @@ public class BundlerUtil {
      */
     public static void objectToXML ( Object obj, File f, boolean removeFirst ) {
 
-        if (xmlSerializer == null) {
-            xmlSerializer = new XStream(new DomDriver("UTF-8"));
-        }
-
         if ( removeFirst && f.exists() )
             f.delete();
 
@@ -253,8 +256,6 @@ public class BundlerUtil {
                 objectToXML(obj, outputStream);
             }
 
-        } catch ( FileNotFoundException e ) {
-            Logger.error( PublisherUtil.class, e.getMessage(), e );
         } catch ( IOException e ) {
             Logger.error( PublisherUtil.class, e.getMessage(), e );
         }
@@ -286,7 +287,7 @@ public class BundlerUtil {
         if ( removeFirst && f.exists() )
             f.delete();
 
-        ObjectMapper mapper = DotObjectMapperProvider.getInstance().getDefaultObjectMapper();
+        final ObjectMapper mapper = getCustomMapper();
 
         try {
             if ( !f.exists() ){
@@ -306,9 +307,23 @@ public class BundlerUtil {
         }
     }
 
+    private static ObjectMapper getObjectMapper() {
+        if (objectMapper == null) {
+            objectMapper = new ObjectMapper();
+        }
+        return objectMapper;
+    }
+
+    private static ObjectMapper getCustomMapper() {
+        if (customMapper == null) {
+            customMapper = DotObjectMapperProvider.getInstance().getDefaultObjectMapper();
+        }
+        return customMapper;
+    }
+
 
     public static void objectToJSON( final Object obj, final OutputStream outputStream) {
-        ObjectMapper mapper = new ObjectMapper();
+        final ObjectMapper mapper = getObjectMapper();
 
         try {
             mapper.writeValue(outputStream, obj);
@@ -333,7 +348,7 @@ public class BundlerUtil {
 	}
 
     public static Object xmlToObject(final InputStream inputStream) throws IOException {
-        XStream xstream = new XStream(new DomDriver("UTF-8"));
+        final XStream xstream = XMLSerializerUtil.getInstance().getXmlSerializer();
 
         try (InputStream input = new BufferedInputStream(inputStream)) {
             return xstream.fromXML(input);
@@ -346,7 +361,7 @@ public class BundlerUtil {
      * @return
      */
     private static Object xmlToObjectWithPrologue(File file) {
-        XStream xstream = new XStream(new DomDriver("UTF-8"));
+        final XStream xstream = XMLSerializerUtil.getInstance().getXmlSerializer();
         XMLUtils.addPrologueIfNeeded(file);
         try (InputStream input = new BufferedInputStream(Files.newInputStream(file.toPath()))) {
             return xstream.fromXML(input);
@@ -368,7 +383,7 @@ public class BundlerUtil {
      * @return A deserialized object
      */
     public static <T> T jsonToObject(File f, Class<T> clazz){
-    	ObjectMapper mapper = new ObjectMapper();
+    	final ObjectMapper mapper = getObjectMapper();
 
     	BufferedInputStream input = null;
 		try {

--- a/dotCMS/src/main/java/com/dotcms/publishing/XMLSerializerUtil.java
+++ b/dotCMS/src/main/java/com/dotcms/publishing/XMLSerializerUtil.java
@@ -5,8 +5,8 @@ import com.thoughtworks.xstream.XStream;
 import com.thoughtworks.xstream.XStreamException;
 import com.thoughtworks.xstream.io.HierarchicalStreamWriter;
 import com.thoughtworks.xstream.io.xml.DomDriver;
-import java.io.OutputStream;
-import java.io.OutputStreamWriter;
+
+import java.nio.charset.StandardCharsets;
 
 /**
  * Util for serialize object to XML file
@@ -14,7 +14,7 @@ import java.io.OutputStreamWriter;
 public class XMLSerializerUtil {
 
     private static final XMLSerializerUtil instance= new XMLSerializerUtil();
-    private XStream xmlSerializer = null;
+    private XStream xmlSerializer;
 
 
     /**
@@ -28,7 +28,11 @@ public class XMLSerializerUtil {
     }
 
     private XMLSerializerUtil(){
-        xmlSerializer = new XStream(new DomDriver("UTF-8"));
+        xmlSerializer = new XStream(new DomDriver(StandardCharsets.UTF_8.name()));
+    }
+
+    public XStream getXmlSerializer() {
+        return xmlSerializer;
     }
 
     /**

--- a/dotCMS/src/main/java/com/dotmarketing/plugin/model/Plugin.java
+++ b/dotCMS/src/main/java/com/dotmarketing/plugin/model/Plugin.java
@@ -1,5 +1,8 @@
 package com.dotmarketing.plugin.model;
 
+import com.dotcms.publisher.util.PusheableAsset;
+import com.dotcms.publishing.manifest.ManifestItem;
+
 import java.io.Serializable;
 import java.util.Date;
 
@@ -9,7 +12,7 @@ import java.util.Date;
  * A simple POJO for dotCMS plugins
  */
 
-public class Plugin implements Serializable{
+public class Plugin implements Serializable, ManifestItem {
 
 	/**
 	 * 
@@ -98,5 +101,13 @@ public class Plugin implements Serializable{
 	public void setLastDeployedDate(Date lastDeployedDate) {
 		this.lastDeployedDate = lastDeployedDate;
 	}
-	
+
+	@Override
+	public ManifestInfo getManifestInfo() {
+		return new ManifestInfoBuilder()
+				.id(id)
+				.title(pluginName)
+				.objectType(PusheableAsset.OSGI.getType())
+				.build();
+	}
 }

